src/hotspot/share/services/nmtPreInit.hpp

```diff
@@ -0,0 +1,353 @@
+
+  // Returns start of the user data area
+  void* payload()             { return this + 1; }
+  const void* payload() const { return this + 1; }
+
+  // These functions do allocate/free a raw malloc block of given payload size,
+  //  preceded with a NMTPreInitAllocation header.
+  static NMTPreInitAllocation* do_alloc(size_t payload_size);
+  static NMTPreInitAllocation* do_reallocate(NMTPreInitAllocation* old, size_t new_payload_size);
+  static void do_free(NMTPreInitAllocation* p);
+};
+
+class NMTPreInitAllocationTable {
+
+  // table_size: keep table size a prime and the hash function simple; this
+  //  seems to give a good distribution for malloce'd pointers on all our libc variants.
+  // 8000ish is really plenty: normal VM runs have ~500 pre-init allocations to hold,
+  //  VMs with insanely long command lines maybe ~700-1000. Which gives us an expected
+  //  load factor of ~.1. Hash collisions should be very rare.
+  // ~8000 entries cost us ~64K for this table (64-bit), which is acceptable.
+  static const int table_size = 7919;
+
+  NMTPreInitAllocation* _entries[table_size];
+
+  typedef int index_t;
+  const index_t invalid_index = -1;
+
+  static unsigned calculate_hash(const void* p) {
+    uintptr_t tmp = p2i(p);
+    unsigned hash = (unsigned)tmp
+                     LP64_ONLY( ^ (unsigned)(tmp >> 32));
+    return hash;
+  }
+
+  static index_t index_for_key(const void* p) {
+    const unsigned hash = calculate_hash(p);
+    return hash % table_size;
+  }
+
+  NMTPreInitAllocation** find_entry(const void* p) const {
+    const unsigned index = index_for_key(p);
+    NMTPreInitAllocation** aa = (NMTPreInitAllocation**) (&(_entries[index]));
+    while ((*aa) != NULL && (*aa)->payload() != p) {
+      aa = &((*aa)->next);
+    }
+    assert((*aa) == NULL || p == (*aa)->payload(),
+           "retrieve mismatch " PTR_FORMAT " vs " PTR_FORMAT ".",
+           p2i(p), p2i((*aa)->payload()));
+    return aa;
+  }
+
+public:
+
+  NMTPreInitAllocationTable();
+
+  // Adds an entry to the table
+  void add(NMTPreInitAllocation* a) {
+    void* payload = a->payload();
+    const unsigned index = index_for_key(payload);
+    assert(a->next == NULL, "entry already in table?");
+    a->next = _entries[index]; // add to front
+    _entries[index] = a;        //   of list
+    assert(find(payload) == a, "add: reverse lookup error?");
+  }
+
+  // Find - but does not remove - an entry in this map.
+  // Returns NULL if not found.
+  NMTPreInitAllocation* find(const void* p) const {
+    NMTPreInitAllocation** aa = find_entry(p);
+    return *aa;
+  }
+
+  // Find and removes an entry from the table. Asserts if not found.
+  NMTPreInitAllocation* find_and_remove(void* p) {
+    NMTPreInitAllocation** aa = find_entry(p);
+    assert((*aa) != NULL, "Entry not found: " PTR_FORMAT, p2i(p));
+    NMTPreInitAllocation* a = (*aa);
+    (*aa) = (*aa)->next;         // remove from its list
```

Comment:
The code is implicitly thread-safe because the hashmap is only modified in the pre-NMT-init phase. After NMT initialization, the table is read-only. During pre-NMT-init we are effectively single-threaded - at most two threads access the map, the thread loading the libjvm, and the thread calling CreateJavaVM, but not at the same time.

See also the asserts in the AllStatic class `NMTPreInit`.

(I should have described it more clearly, will add a comment.)

Reply:
So, you are saying that there is no memory that is malloc'd pre-NMT-init phase and freed post-NMT-init phase?

Reply:
Okay, I see. It just leaks those memory, so the table can be read-only.

Reply:
Exaclty. There are a few allocs which either get free'd or realloc'ed post-init, but not enough to make freeing them worth it if it means having to serialize access to the lookup table.